alacritty/res/rect.f.glsl

```diff
@@ -9,46 +9,120 @@ flat in vec4 color;
+  float undercurlTop = undercurl + max((underlineThickness - 1), 0);
+  float undercurlBottom = undercurl - max((underlineThickness - 1), 0);
 
-    // We use `undercurlPosition` as amplitude, since it's half of the descent
-    // value.
-    float undercurl = -1. * undercurlPosition / 2.
-        * cos(float(x) * 2 * PI / float(cellWidth))
-        + cellHeight - undercurlPosition;
+  // Compute resulted alpha based on distance from `gl_FragCoord.y` to the
+  // cosine curve.
+  float alpha = 1.;
+  if (y > undercurlTop || y < undercurlBottom) {
+    alpha = 1. - min(abs(undercurlTop - y), abs(undercurlBottom - y));
+  }
 
-    // We subtract one, since curve is already 1px thick.
-    float undercurl_top = undercurl + max((undercurlThickness - 1), 0);
-    float undercurl_bottom = undercurl - max((undercurlThickness - 1), 0);
+  // The result is an alpha mask on a rect, which leaves only curve opaque.
+  return vec4(color.rgb, alpha);
+}
 
+// When the dot size increases we can use AA to make spacing look even and the
+// dots rounded when the size allows us.
+vec4 draw_dotted_aliased(float x, float y) {
+  int dotNumber = int(x / underlineThickness);
 
-    // Compute resulted alpha based on distance from `gl_FragCoord.y` to the
-    // cosine curve.
-    float alpha = 1.;
-    if (y > undercurl_top || y < undercurl_bottom) {
-        alpha = 1. - min(abs(undercurl_top - y), abs(undercurl_bottom - y));
-    }
+  float radius = underlineThickness / 2.;
+  float centerY = cellHeight - underlinePosition;
+
+  float leftCenter = (dotNumber - dotNumber % 2) * underlineThickness + radius;
+  float rightCenter = leftCenter + 2 * underlineThickness;
+
+  float distanceLeft = sqrt(pow(x - leftCenter, 2) + pow(y - centerY, 2));
+  float distanceRight = sqrt(pow(x - rightCenter, 2) + pow(y - centerY, 2));
+
+  float alpha = max(1 - (min(distanceLeft, distanceRight) - radius), 0);
+  return vec4(color.rgb, alpha);
+}
+
+/// Draw dotted line when dot is just a single pixel.
+vec4 draw_dotted(int x, int y) {
+  int cellEven = 0;
+
+  // To make spacing evenly spread our dot and its gap should divide
+  // cellWidth, however since the dot + gap is 2 for 1 px dot we should use 2 *
+  // cellWidth when cellWidth is odd.
+  if (int(cellWidth) % 2 != 0) {
+    cellEven = int((gl_FragCoord.x - paddingX) / cellWidth) % 2;
+  }
 
-    // The result is an alpha mask on a rect, which leaves only curve opaque.
-    FragColor = vec4(color.xyz, alpha);
+  // Since we're using the entire descent area for dots as well we should
+  // compute `y` distance to them, so won't look like bars.
+  float alpha = 1. - abs(round(cellHeight - underlinePosition) - y);
+  if (x % 2 != cellEven) {
+    alpha = 0;
+  }
+
+  return vec4(color.rgb, alpha);
+}
+
+vec4 draw_dashed(int x) {
+  // Compute the half of the gap, since on the dash is in the middle of the
+  // cell and half_gap from edges, so when we'll have adjacent dashed cell
+  // gaps will have the same size as dash.
```

Comment:
This comment seem somewhat convoluted I'm not quite sure what you're trying to say but it might be beneficial to keep it a bit shorter and more precise?